sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/AliasAwareOutputExpression.scala

```diff
@@ -0,0 +1,113 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.sql.catalyst.SQLConfHelper
+import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, Empty2Null, Expression, NamedExpression, SortOrder}
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * A trait that provides functionality to handle aliases in the `outputExpressions`.
+ */
+trait AliasAwareOutputExpression[T <: QueryPlan[T]]
+  extends SQLConfHelper { self: QueryPlan[T] =>
+  private val aliasCandidateLimit = conf.getConf(SQLConf.OUTPUT_EXPRESSION_ALIAS_CANDIDATE_LIMIT)
+  protected def outputExpressions: Seq[NamedExpression]
+  /**
+   * This method can be used to strip expression which does not affect the result, for example:
+   * strip the expression which is ordering Agnostic for output ordering.
+   */
+  protected def strip(expr: Expression): Expression = expr
+
+  lazy val aliasMap = {
+    val attrWithAliasMap = new mutable.HashMap[Expression, ArrayBuffer[Attribute]]()
+
+    def updateAttrWithAliasMap(key: Expression, target: Attribute): Unit = {
+      val aliasArray = attrWithAliasMap.getOrElseUpdate(strip(key).canonicalized,
+        new ArrayBuffer[Attribute]())
+      // pre-filter if the number of alias has bigger than candidate limit
+      if (aliasArray.size < aliasCandidateLimit) {
+        aliasArray.append(target)
+      }
+    }
+
+    outputExpressions.foreach {
+      case a @ Alias(child, _) =>
+        updateAttrWithAliasMap(child, a.toAttribute)
+      case _ =>
+    }
+    attrWithAliasMap.toMap
+  }
+
+  protected def hasAlias: Boolean = aliasMap.nonEmpty
+
+  // Return a set of Expression which normalize the original expression to the aliased.
+  protected def normalizeExpression(expr: Expression): Seq[Expression] = {
+    if (aliasCandidateLimit < 1) {
+      expr :: Nil
+    } else {
+      val normalizedCandidates = new ArrayBuffer[Expression]()
+      // Stop loop if the size of candidates exceed limit
+      for ((origin, aliases) <- aliasMap if normalizedCandidates.size < aliasCandidateLimit) {
+        for (alias <- aliases if normalizedCandidates.size < aliasCandidateLimit) {
+          val newExpr = expr.transformDown {
+            case e: Expression if e.canonicalized == origin => alias
+          }
+          if (!expr.fastEquals(newExpr)) {
+            normalizedCandidates.append(newExpr)
```

Comment:
We should only append the new expr if all its references are contained by `outputExpressions.map(_.toAttribute)`

Reply:
newExpr can contain other reference. for example `df.orderby($"a" + $"b").selectExpr("a as x")`, we only replace `a` to `x` but the expression `Add` has an another attribute `b`.

Reply:
in this case, we should not report the ordering as `x + b`, as `b` is not even outputted by the plan.

Reply:
then how about `df.orderby($"a" + $"b").selectExpr("a as x", "b")` ?

Reply:
it's a little hard do this in `AliasAwareOutputExpression`. for example  `PartitioningCollection(a, b)` and a alias to x. If we want to return PartitioningCollection(a) only, then we need to prune b. It should be handled at `AliasAwareOutputPartitioning`?

Reply:
We can't return `PartitioningCollection(a)` only. If a relation's child is partitioned by `a` and `b`, but `b` is not outputted by this relation (no alias either), then it's wrong to say this relation is partitioned by `a`. It can only be `UnknownPartitioning`.

Reply:
`PartitioningCollection(a, b)` means `t1 join t2 on a = b`, not `group by a, b`..

Reply:
Ah, then it's a flatMap semantic